src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -499,13 +500,36 @@ sliceSepByMax p n = sliceSepBy p . FL.ltake n
 
 -- | See 'Streamly.Internal.Data.Parser.wordBy'.
 --
--- /Unimplemented/
 --
-{-# INLINABLE wordBy #-}
-wordBy ::
-    -- Monad m =>
-    (a -> Bool) -> Fold m a b -> Parser m a b
-wordBy = undefined
+{-# INLINE wordBy #-}
+wordBy :: Monad m => (a -> Bool) -> Fold m a b -> Parser m a b
+wordBy predicate (Fold fstep finitial fextract) = Parser step initial extract
+
+    where
+
+    {-# INLINE worder #-}
+    worder s a = do
+        res <- fstep s a
+        return
+            $ case res of
+                  FL.Partial s1 -> Partial 0 $ Right' s1
+                  FL.Done b -> Done 0 b
+
+    initial = Left' <$> finitial
+
+    step (Left' s) a =
+        if not (predicate a)
+        then worder s a
+        else return $ Partial 0 $ Left' s
+    step (Right' s) a =
+        if not (predicate a)
+        then worder s a
+        else do
+            ex <- fextract s
+            return $ Done 1 $ ex
```

Comment:
Since we are finishing right after the word ends, as I suggested earlier, this will result into an empty word being emitted in the end if there is white space in the end.

We need a test case that when this parser is used on a stream of a word surrounded by whitespace e.g. " hello " results in a single word being emitted i.e. ["hello"].

For short scopes, we can just use `r` for a temporary result instead of using different names every time.